synapse/cryotank.py

```diff
@@ -571,3 +588,609 @@ def init(self, name, conf=None, timeout=None):
+        self.indices = {k: _MetaEntry(**v) for k, v in indices.get('present', {}).items()}
+        self.deleting = list(indices.get('deleting', ()))
+        # Keeps track (non-persistently) of which indicies have been paused
+        self.asleep = defaultdict(bool)  # type: ignore
+
+        # How far each index has progressed as well as statistics
+        self.progresses = s_msgpack.un(progress_enc)
+        if not all(p in self.indices for p in self.deleting):
+            raise s_exc.CorruptDatabase('index meta table: deleting entry with unrecognized property name')
+        if not all(p in self.indices for p in self.progresses):
+            raise s_exc.CorruptDatabase('index meta table: progress entry with unrecognized property name')
+        if is_new_db:
+            self.persist()
+
+    def persist(self, progressonly=False, txn=None):
+        '''
+        Persists the index info to the database
+
+        Args:
+            progressonly (bool): if True, only persists the progress (i.e. more dynamic) information
+            txn (Optional[lmdb.Transaction]): if not None, will use that transaction to record data.  txn is
+            not committed.
+
+        Returns:
+            None
+        '''
+        d = {'delete': self.deleting,
+             'present': {k: v._asdict() for k, v in self.indices.items()}}
+
+        with contextlib.ExitStack() as stack:
+            if txn is None:
+                txn = stack.enter_context(self._dbenv.begin(db=self._metatbl, buffers=True, write=True))
+            if not progressonly:
+                txn.put(b'indices', s_msgpack.en(d))
+            txn.put(b'progress', s_msgpack.en(self.progresses))
+
+    def lowestProgress(self):
+        '''
+        Returns:
+            int: The next offset that should be indexed, based on active indices.
+        '''
+        nextoffsets = [p['nextoffset'] for iid, p in self.progresses.items() if not self.asleep[iid]]
+        return min(nextoffsets) if nextoffsets else s_lmdb.MAX_INT_VAL
+
+    def iidFromProp(self, prop):
+        '''
+        Returns:
+            int: the index id for the propname, None if not found
+        '''
+        return next((k for k, idx in self.indices.items() if idx.propname == prop), None)
+
+    def addIndex(self, prop, syntype, datapath, *args):
+        '''
+        Adds an index to the cryotank.
+
+        Args:
+            prop (str):  the name of the property this will be stored as in the normalized record
+            syntype (str):  the synapse type this will be interpreted as
+            datapath (str):  the datapath spec against which the raw record is run to extract a single field
+        that is passed to the type normalizer.
+            *args (str):  additional datapaths that will be tried in order if the first isn't present.
+        Returns:
+            None
+
+        N.B.  additional datapaths will be tried iff prior datapaths are not present, and *not* if
```

Comment:
Should this be in the ``Notes`` section of the docstring?

Reply:
Well, N.B. is nota bene, so yes.